WhateverGreen/kern_igfx_memory.cpp

```diff
@@ -0,0 +1,199 @@
+//
+//  kern_igfx_memory.cpp
+//  WhateverGreen
+//
+//  Created by FireWolf on 8/30/20.
+//  Copyright © 2020 vit9696. All rights reserved.
+//
+
+#include "kern_igfx.hpp"
+#include <Headers/kern_util.hpp>
+#include <Headers/kern_disasm.hpp>
+
+void IGFX::DVMTCalcFix::init() {
+	// We only need to patch the framebuffer driver
+	requiresPatchingGraphics = false;
+	requiresPatchingFramebuffer = true;
+}
+
+void IGFX::DVMTCalcFix::deinit() {}
+
+void IGFX::DVMTCalcFix::processKernel(KernelPatcher &patcher, DeviceInfo *info) {
+	// Guard: Disable the patch if it is not available on the current Intel platforms
+	if (!available) {
+		SYSLOG("igfx", "DVMT: This fix is not available on the current platform and has been disabled.");
+		return;
+	}
+		
+	// Enable the fix if designated boot argument or device property is found
+	enabled = checkKernelArgument("-igfxdvmt");
+	if (!enabled)
+		enabled = info->videoBuiltin->getProperty("enable-dvmt-calc-fix") != nullptr;
+	
+	// Read the DVMT preallocated memory set in BIOS from the GMCH Graphics Control field at 0x50 (PCI0,2,0)
+	// TODO: Lilu needs to be updated to define the enumeration case `kIOPCIConfigGraphicsControl`
+	auto gms = WIOKit::readPCIConfigValue(info->videoBuiltin, /*WIOKit::kIOPCIConfigGraphicsControl*/ 0x50, 0, 16) >> 8;
+	
+	// Disable the fix if the GMS value can be calculated by Apple's formula correctly
+	// Reference: 10th Generation Intel Processor Families: Datasheet, Volume 2, Section 4.1.28
+	// https://www.intel.com/content/www/us/en/products/docs/processors/core/10th-gen-core-families-datasheet-vol-2.html
+	if (gms < 0x10) {
+		dvmt = gms * 32;
+		enabled = false;
+		DBGLOG("igfx", "DVMT: GMS value is supported by the driver. The fix has been disabled.");
+	} else if (gms == 0x20 || gms == 0x30 || gms == 0x40) {
+		dvmt = gms * 32;
+	} else if (gms >= 0xF0 && gms <= 0xFE) {
+		dvmt = ((gms & 0x0F) + 1) * 4;
+	} else {
+		SYSLOG("igfx", "DVMT: GMS value is reserved. Check your BIOS settings. DVMT will be set to 0.");
+		return;
+	}
+	
+	DBGLOG("igfx", "DVMT: GMS value is 0x%02x; DVMT pre-allocated memory is %d MB.", gms, dvmt);
+	info->videoBuiltin->setProperty("fw-dvmt-gms-field-value", gms, 8);
```

Comment:
Is this for debugging? If so, guard it with ifdef DEBUG maybe?

Reply:
Not exactly for debugging, but I think it's better to make it only available in DEBUG version.